### PR TITLE
Remove RSA padding SSLv23

### PR DESCRIPTION
**Pull request by richsalz**
Fixes: #14216

I am not sure how to handle some of the documentation updates, particularly `man7/EVP_SIGNATURE-RSA.pod` Just mark it inline as deprecated? I kinda did that for `man3/EVP_PKEY_CTX_ctrl.pod`

### CODE REVIEW

File: crypto/rsa/build.info
Comment (t8m) [inline crypto/rsa/build.info:12]: It should be placed into !$disabled{'deprecated-3.0'} if down there instead.

File: doc/man3/EVP_PKEY_CTX_ctrl.pod
Comment (t8m) [inline doc/man3/EVP_PKEY_CTX_ctrl.pod:246]: Please keep it here, just add deprecated. `deprecated B<RSA_SSLV23_PADDING> for SSLv23 padding`

File: include/openssl/core_names.h
Comment (t8m) [inline include/openssl/core_names.h:387]: Maybe we should just drop this define? So we do not add immediately deprecated defines in 3.0?

**Comment by t8m** [comment]
> I am not sure how to handle some of the documentation updates, particularly `man7/EVP_SIGNATURE-RSA.pod` Just mark it inline as deprecated? I kinda did that for `man3/EVP_PKEY_CTX_ctrl.pod`

I'd just drop it from the document and kept it only as an internal implementation detail that there is another possible value for the pad mode. It is really useless and I would not expect any third party provider to implement this.

File: crypto/rsa/build.info
Comment (richsalz) [reply crypto/rsa/build.info:12]: yeah, oops.  fixed in the updated commit.

File: doc/man3/EVP_PKEY_CTX_ctrl.pod
Comment (richsalz) [reply doc/man3/EVP_PKEY_CTX_ctrl.pod:246]: I put it as `(deprecated)` to follow the structure of the other options.   Whew, that's a long sentence, should probably be turned into a bullet-list but that's for another PR.

File: include/openssl/core_names.h
Comment (richsalz) [reply include/openssl/core_names.h:387]: I agree adding something and deprecating it seems silly, but it's used in two files so have to keep it I think.

**Comment by richsalz** [comment]
Once again I can't seem to get no-deprecated correct.  help please.

File: include/openssl/core_names.h
Comment (t8m) [reply include/openssl/core_names.h:387]: Could it be defined in some internal header? @levitte any opinions?

File: crypto/rsa/rsa_ossl.c
Comment (t8m) [inline crypto/rsa/rsa_ossl.c:122]: This change has no effect because you have the `#include "internal/deprecated.h"` above. Unfortunately that means the rsa_ssl.c cannot be among the `!$disabled{'deprecated-3.0'}` files.

Comment (t8m) [reply crypto/rsa/rsa_ossl.c:122]: This is the cause for the no-deprecated build failure.

Comment (richsalz) [reply crypto/rsa/rsa_ossl.c:122]: `rsa_ossl` needs to use deprecated stuff and `rsa_ssl` implements deprecated stuff. So it seems that using the standard deprecation flags will not work.  Does anyone have suggestions on how to do this deprecation?  One possibility I am about to try is to add a shim file that doesn't include `internal/deprecated.h` Let's see what happens.

Comment (t8m) [reply crypto/rsa/rsa_ossl.c:122]: No, you're misunderstanding. The fix is just to drop the move of rsa_ssl.c into the `!$disabled{'deprecated-3.0'}` block. Just drop the build.info change.

Comment (richsalz) [reply crypto/rsa/rsa_ossl.c:122]: > Just drop the build.info change.

well heck, that's easier. :)

**Comment by richsalz** [comment]
Dang I really twisted up my repo.  sigh.  closing this for now.

File: doc/man1/openssl-pkeyutl.pod.in
Comment (mcepl) [inline doc/man1/openssl-pkeyutl.pod.in:239]: Typo: deprecaed -> deprecated.

Comment (richsalz) [reply doc/man1/openssl-pkeyutl.pod.in]: Thanks @mcepl ! Updated commit pushed.

**Comment by t8m** [comment]
Now, thinking about this more, and I am sorry for not proposing this earlier, I'd say this should be removed completely and not just deprecated. However that needs an OMC (or OTC?) approval.

**Comment by richsalz** [comment]
This is on hold pending OTC decision in #14283. And if that is approved, I'd be happy to update this.

**Comment by t8m** [comment]
The vote for removal has passed. So the RSA_SSLV23_PADDING and any related things should be completely removed.

**Comment by t8m** [comment]
I forgot we have an issue for the removal already. So closing this one as it is duplicate.

**Comment by t8m** [comment]
I really need to get some coffee break - this one is a PR and not an issue. So @richsalz do you want to morph this into removal PR? Or do you want to open a new PR for the removal? Or, if you do not want to or have time to work on the removal, I can as well do it.

**Comment by richsalz** [comment]
I had it lying around in anticipation :)  I only documented in CHANGES, let me know if that's not okay.

**Comment by t8m** [comment]
You'll need to rebase against master and fix the remaining places.

File: util/libcrypto.num
Comment (t8m) [inline util/libcrypto.num:1395]: Not sure if those should not be rather NOEXIST::FUNCTION entries as these are numbered entries and similarly below. IMO make update would do that instead of deletion.

**Comment by richsalz** [comment]
Yeah, I missed one line; updated commit pushed.  Based on current master.   FYI the edits to `rsautl.pod` removed some duplicated lines.

File: CHANGES.md
Comment (t8m) [inline CHANGES.md:81]: Maybe also mention that the `rsautl -ssl` is removed and those two public functions from the RSA API are removed as well.

**Comment by richsalz** [comment]
fixup commit pushed address @t8m 's feedback.

File: CHANGES.md
Comment (t8m) [inline CHANGES.md:81]: Should it be rather `includes`? And please use backticks around the `-ssl` and `rsautl`.

Comment (richsalz) [reply CHANGES.md]: updated commit pushed.

File: CHANGES.md
Comment (t8m) [inline CHANGES.md:82]: Oops there is also a spurious `-` before RSA_padding_add_SSLv23.

**Review by t8m — APPROVED** [review]
Approving provided the last nit is fixed as well.

File: CHANGES.md
Comment (richsalz) [reply CHANGES.md]: updated fixup commit pushed.

(four word-sentence, 3 of them jargon :)

File: doc/man7/provider-asym_cipher.pod
Comment (kroeckx) [inline doc/man7/provider-asym_cipher.pod:186]: Should we renumber?

**Review by kroeckx — APPROVED** [review]

File: doc/man7/provider-asym_cipher.pod
Comment (richsalz) [reply doc/man7/provider-asym_cipher.pod:186]: I suggest not renumbering, since an old binary with a new shared library will fail with "unknown padding" rather than silently failing to do the expected thing.

**Comment by mcepl** [comment]
For your enjoyment, https://gitlab.com/m2crypto/m2crypto/-/merge_requests/264

File: doc/man7/provider-asym_cipher.pod
Comment (kroeckx) [reply doc/man7/provider-asym_cipher.pod:186]: Since this targets 3.0 only, there can be no old binaries for this shared library yet. But keeping them is fine for me.

Comment (t8m) [reply doc/man7/provider-asym_cipher.pod:186]: There are plenty of numbers so I do not see a real reason for renumbering.

**Comment by t8m** [comment]
Merged to master. Thank you for the contribution.

**Pull request closed without merging**
